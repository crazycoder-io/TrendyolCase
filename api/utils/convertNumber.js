module.exports = (value) => `${value.toString().split(".")[0]}.${value.toString().split(".")[1].toString().substring(0, 3)}`;
