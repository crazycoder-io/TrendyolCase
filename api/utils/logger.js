"use strict";

const chalk = require("chalk");
const morgan = require("morgan");

morgan.token("id", (req) => req.id.split("-")[0]);

module.exports = {
    requestHandleLogger: morgan(
        chalk.hex("#fffa65").bold("ğŸ“© --> [:date[iso] #:id] Started :method :url for :remote-addr"),
        { immediate: true },
    ),
    responseHandleLogger: morgan(chalk.hex("#2ed573").bold("âœ… --> [:date[iso] #:id] Completed :status :res[content-length] in :response-time ms:")),
};
