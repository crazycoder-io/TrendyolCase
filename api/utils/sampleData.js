module.exports = {
    SAMPLE_DATA_1: {
        ttfb: 0.191,
        domLoad: 15.556,
        windowLoad: 15.556,
        fcp: 13.238,
        resources: [
          {
            name: "https://fonts.googleapis.com/css2?family=Rubik:ital,wght@1,300&display=swap",
            initiatorType: link,
            responseEnd: 13.056,
            transferSize: 1176
          },
          {
            name: "https://fonts.googleapis.com/css2?family=Karla:wght@300&display=swap",
            initiatorType: link,
            responseEnd: 13.128,
            transferSize: 0
          },
          {
            name: "https://perf-analytics-client.herokuapp.com/client/css/styles.css",
            initiatorType: link,
            responseEnd: 13.104,
            transferSize: 1862
          },
          {
            name: "https://perf-analytics-client.herokuapp.com/client/js/perfAnalytics.min.js",
            initiatorType: script,
            responseEnd: 13.429,
            transferSize: 1719
          },
          {
            name: "https://blog.crazycoder.io/images/portfolio-pp.jpg",
            initiatorType: img,
            responseEnd: 14.387,
            transferSize: 0
          },
          {
            name: "https://www.youtube.com/embed/soL62aIlqoM",
            initiatorType: iframe,
            responseEnd: 13.191,
            transferSize: 0
          },
          {
            name: "https://perf-analytics-client.herokuapp.com/favicon.ico",
            initiatorType: other,
            responseEnd: 14.258,
            transferSize: 0
          },
          {
            name: "https://fonts.gstatic.com/s/rubik/v14/iJWbBXyIfDnIV7nEt3KSJbVDV49rz8sDE3U3f4I.woff2",
            initiatorType: other,
            responseEnd: 13.926,
            transferSize: 0
          },
          {
            name: "https://fonts.gstatic.com/s/karla/v15/qkBIXvYC6trAT55ZBi1ueQVIjQTDppqaE0lK.woff2",
            initiatorType: other,
            responseEnd: 13.5,
            transferSize: 12097
          },
          {
            name: "https://fonts.gstatic.com/s/karla/v15/qkBIXvYC6trAT55ZBi1ueQVIjQTDppqaHUlKd7c.woff2",
            initiatorType: other,
            responseEnd: 13.48,
            transferSize: 5684
          }
        ],
        UserAgent: "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:90.0) Gecko/20100101 Firefox/90.0",
        URL: "https://perf-analytics-client.herokuapp.com/"
    }
    
};