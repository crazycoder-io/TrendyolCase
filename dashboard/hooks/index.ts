import { useState, useCallback } from "react";
import { ChartsData, _Date } from "../types";

export const LoadingHook = (): { loading: boolean, updateLoading: () => void } => {
    const [loading, setLoading] = useState<boolean>(false);

    const updateLoading = useCallback(() => setLoading((prev) => !prev), []);

    return { loading, updateLoading };
};

export const ChartHook = (charts?: ChartsData): { chartsData: ChartsData, updateChartsData: (data) => void } => {
    const [chartsData, setChartsData] = useState<ChartsData>(charts);

    const updateChartsData = useCallback((updatedCharts: ChartsData) => setChartsData(updatedCharts), []);

    return { chartsData, updateChartsData };
};

export const DateHook = (): { specificDate: _Date, updateDate: (key, data) => void } => {
    const [specificDate, setDate] = useState<_Date>({
        startDate: new Date(),
        endDate: new Date(),
    });

    const updateDate = useCallback((keyword, date) => setDate(prev => ({...prev, [keyword]: date})), [specificDate]);

    return { specificDate, updateDate };
};
