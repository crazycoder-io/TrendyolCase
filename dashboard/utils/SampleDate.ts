import { ChartsData, Metrics } from "../types";

export const CHARTS_DATA: ChartsData = {
  ttfb: [
    {
      "x": "2:22:51 AM",
      "y": 0.551
    },
    {
      "x": "2:22:52 AM",
      "y": 0.287,
    },
    {
      "x": "2:22:53 AM",
      "y": 0.563,
    },
    {
      "x": "2:22:54 AM",
      "y": 0.101,
    }
  ],
  
  domLoad: [
    {
      "x": "2:22:51 AM",
      "y": 2.045
    },
    {
      "x": "2:22:52 AM",
      "y": 0.456
    },
    {
      "x": "2:22:53 AM",
      "y": 2.567
    },
    {
      "x": "2:22:54 AM",
      "y": 1.896
    }
  ],
  
  fcp: [
    {
      "x": "2:22:51 AM",
      "y": 1.3452000000029801
    },
    {
      "x": "2:22:52 AM",
      "y": 0.1234000000059605
    },
    {
      "x": "2:22:53 AM",
      "y": 0.911
    },
    {
      "x": "2:22:54 AM",
      "y": 1.563615000002901
    }
  ],
  
  windowLoad: [
    {
      "x": "2:22:51 AM",
      "y": 2.456
    },
    {
      "x": "2:22:52 AM",
      "y": 1.23
    },
    {
      "x": "2:22:53 AM",
      "y": 2.101
    },
    {
      "x": "2:22:54 AM",
      "y": 1.989
    }
  ]
};

export const METRICS: Metrics = [{
  "resources": [
    {
      "name": "https://fonts.googleapis.com/css2?family=Rubik:ital,wght@1,300&display=swap",
      "initiatorType": "link",
      "responseEnd": 0.6699000000059605,
      "transferSize": 0
    },
    {
      "name": "https://fonts.googleapis.com/css2?family=Karla:wght@300&display=swap",
      "initiatorType": "link",
      "responseEnd": 0.8762000000029803,
      "transferSize": 0
    },
    {
      "name": "https://perf-analytics-client.herokuapp.com/client/css/styles.css",
      "initiatorType": "link",
      "responseEnd": 1.0439000000059604,
      "transferSize": 1909
    },
    {
      "name": "https://blog.crazycoder.io/images/portfolio-pp.jpg",
      "initiatorType": "img",
      "responseEnd": 1.8305,
      "transferSize": 0
    },
    {
      "name": "https://perf-analytics-client.herokuapp.com/client/js/perfAnalytics.min.js",
      "initiatorType": "script",
      "responseEnd": 0.8561000000014901,
      "transferSize": 2205
    },
    {
      "name": "https://fonts.googleapis.com/css2?family=Rubik:ital,wght@1,300&display=swap",
      "initiatorType": "other",
      "responseEnd": 0.9022000000029803,
      "transferSize": 778
    },
    {
      "name": "https://fonts.googleapis.com/css2?family=Karla:wght@300&display=swap",
      "initiatorType": "other",
      "responseEnd": 0.9665,
      "transferSize": 658
    },
    {
      "name": "https://fonts.gstatic.com/s/rubik/v14/iJWbBXyIfDnIV7nEt3KSJbVDV49rz8sDE3U3f4TnlY1PK6w.woff",
      "initiatorType": "css",
      "responseEnd": 1.2192000000029801,
      "transferSize": 0
    },
    {
      "name": "https://fonts.gstatic.com/s/karla/v15/qkBIXvYC6trAT55ZBi1ueQVIjQTDppqaE0lMZbLXGhmR.woff",
      "initiatorType": "css",
      "responseEnd": 1.2111000000014902,
      "transferSize": 0
    },
    {
      "name": "https://fonts.gstatic.com/s/karla/v15/qkBIXvYC6trAT55ZBi1ueQVIjQTDppqaHUlMZbLXGhmRytc.woff",
      "initiatorType": "css",
      "responseEnd": 1.1963000000044703,
      "transferSize": 0
    }
  ],
  "createdAt": new Date("2021-08-08T23:22:51.219Z"),
  "_id": "611067f680f18100035647e1",
  "ttfb": 0.351,
  "domLoad": 1.045,
  "windowLoad": 2.078,
  "fcp": 1.1162000000029801,
  "UserAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.131 Safari/537.36",
  "URL": "https://perf-analytics-client.herokuapp.com/",
  "__v": 0
}];
