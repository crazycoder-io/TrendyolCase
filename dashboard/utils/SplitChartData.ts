import moment from "moment";
import "moment/locale/tr";
moment.locale("tr");
import { Metrics, ChartsData } from "../types";
const SplittedData: ChartsData = {
    ttfb: [],
    fcp: [],
    domLoad: [],
    windowLoad: [],
};

const SplitData = (metrics: Metrics): ChartsData => {
    metrics.map((metric) => {
        SplittedData.ttfb.push({
            x: moment(metric.createdAt).format("DD/MM HH:mm").toString(),
            y: metric.ttfb,
        });
    
        SplittedData.fcp.push({
            x: moment(metric.createdAt).format("DD/MM HH:mm").toString(),
            y: metric.fcp,
        });
    
        SplittedData.domLoad.push({
            x: moment(metric.createdAt).format("DD/MM HH:mm").toString(),
            y: metric.domLoad,
        });
    
        SplittedData.windowLoad.push({
            x: moment(metric.createdAt).format("DD/MM HH:mm").toString(),
            y: metric.windowLoad,
        });
    });
    return SplittedData;
};

export default SplitData;