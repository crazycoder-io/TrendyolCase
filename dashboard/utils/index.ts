export { default as SplitChartData } from "./SplitChartData";