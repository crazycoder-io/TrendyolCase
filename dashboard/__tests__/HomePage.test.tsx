import { shallow } from "enzyme";
import { act, renderHook } from "@testing-library/react-hooks";
import { ChartHook } from "../hooks";
import { CHARTS_DATA, METRICS } from "../utils/SampleDate";
import IndexPage from "../pages/index";

describe("Index page unit test", () => {
    let wrapper;

    beforeAll(() => {
        wrapper = shallow(<IndexPage chartsData={CHARTS_DATA} metrics={METRICS} />);
    });
    test("Index page components should be defined well", () => {
        
        expect(wrapper).toMatchSnapshot();

        const HeaderComponent = wrapper.find("Header");
        const TitleComponent = wrapper.find("Title");
        const ChartsComponent = wrapper.find("Charts");
        const FooterComponent = wrapper.find("Footer");
        const DatePickerComponent = wrapper.find("DatePicker");
        
        expect(HeaderComponent).toBeDefined();
        expect(TitleComponent).toBeDefined();
        expect(ChartsComponent).toBeDefined();
        expect(FooterComponent).toBeDefined(); 
        expect(DatePickerComponent).toBeDefined(); 
    });

    test("Mocking 'Update Charts' function", () => {
        const { result } = renderHook(() => ChartHook(CHARTS_DATA));
        const myMockFn = jest
            .fn()
            .mockImplementationOnce(result.current.updateChartsData);
        
        act(() => myMockFn());
        expect(myMockFn).toHaveBeenCalledTimes(1);
    });

    test("'Update Charts' button simulate 'click' without error", () => {
        const UpdateChartsButton = wrapper.find(".dateButton");
        ;
        UpdateChartsButton.simulate("click");
    });
});