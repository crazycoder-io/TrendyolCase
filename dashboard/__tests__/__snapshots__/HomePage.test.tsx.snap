// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page unit test Index page components should be defined well 1`] = `ShallowWrapper {}`;
